private/buf/bufsync/syncer.go

```diff
@@ -280,217 +309,177 @@ func (s *syncer) validateDefaultBranch(ctx context.Context, moduleIdentity bufmo
-			if !shouldSyncModule {
-				s.logger.Debug(
-					"module directory not present as a module to sync, skipping module in commit",
-					zap.String("branch", branch),
-					zap.String("commit", commitHash),
-					zap.String("module directory", moduleDir),
-				)
-				continue
-			}
-			if builtModule == nil {
-				s.logger.Debug(
-					"module directory has no module to sync, skipping module in commit",
-					zap.String("branch", branch),
-					zap.String("commit", commitHash),
-					zap.String("module directory", moduleDir),
-				)
-				continue
-			}
-			modIdentity := builtModule.ModuleIdentity().IdentityString()
-			if err := syncFunc(
-				ctx,
-				newModuleCommit(
-					branch,
-					commitForSync.commit,
-					s.commitsToTags[commitHash],
-					moduleDir,
-					builtModule.ModuleIdentity(),
-					builtModule.Bucket,
-				),
-			); err != nil {
-				return fmt.Errorf("sync module %s:%s in commit %s: %w", moduleDir, modIdentity, commitHash, err)
-			}
-			// module was synced successfully, add it to the cache
-			if s.modulesIdentitiesToCommitsSyncedCache[modIdentity] == nil {
-				s.modulesIdentitiesToCommitsSyncedCache[modIdentity] = make(map[string]struct{})
-			}
-			s.modulesIdentitiesToCommitsSyncedCache[modIdentity][commitHash] = struct{}{}
+		if builtModule.ModuleIdentity() == nil {
+			return fmt.Errorf("built module for commit %s has no module identity", commitHash)
 		}
+		modIdentity := builtModule.ModuleIdentity().IdentityString()
+		if err := syncFunc(
+			ctx,
+			newModuleCommit(
+				branch,
+				commitForSync.commit,
+				s.commitsToTags[commitHash],
+				moduleDir,
+				builtModule.ModuleIdentity(),
+				builtModule.Bucket,
+			),
+		); err != nil {
+			return fmt.Errorf("sync module %s:%s in commit %s: %w", moduleDir, modIdentity, commitHash, err)
+		}
+		// module was synced successfully, add it to the cache
+		if s.modulesIdentitiesToCommitsSyncedCache[modIdentity] == nil {
+			s.modulesIdentitiesToCommitsSyncedCache[modIdentity] = make(map[string]struct{})
+		}
+		s.modulesIdentitiesToCommitsSyncedCache[modIdentity][commitHash] = struct{}{}
 	}
 	return nil
 }
 
-// branchSyncableCommits returns a sorted commit+modules tuples array that are pending to sync for a
-// branch. A commit in the array might have no modules to sync if those are skipped by the Syncer
-// error handler, or are a found sync point.
-func (s *syncer) branchSyncableCommits(ctx context.Context, branch string) ([]*syncableCommit, error) {
-	branchModulesForSync, ok := s.branchesToModulesForSync[branch]
-	if !ok || len(branchModulesForSync) == 0 {
-		// branch should not be synced, or no modules to sync in that branch
-		return nil, nil
-	}
-	// Copy all branch modules to sync and mark them as pending, until its starting sync point is
-	// reached. They'll be removed from this list as its initial sync point is found.
-	pendingModules := s.copyBranchModulesSync(branch, branchModulesForSync)
```

Comment:
No need for the concept of "pending modules" anymore, this func is invoked once per module+branch, so each time it's just a single module.